File: src/main/java/play/pluv/feed/controller/FeedController.java
package play.pluv.feed.controller;

import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import play.pluv.base.BaseResponse;
import play.pluv.feed.application.FeedService;
import play.pluv.feed.application.dto.FeedDetailResponse;
import play.pluv.feed.application.dto.FeedListResponse;
import play.pluv.security.JwtMemberId;

@RestController
@RequiredArgsConstructor
@RequestMapping("/feed")
public class FeedController {

  private final FeedService feedService;

  @GetMapping
  public BaseResponse<List<FeedListResponse>> getFeeds() {
    final var feeds = feedService.findAll();
    final List<FeedListResponse> responses = FeedListResponse.createList(feeds);
    return BaseResponse.ok(responses);
  }

  @GetMapping("/{id}")
  public BaseResponse<FeedDetailResponse> getFeed(
      @PathVariable final Long id, final JwtMemberId jwtMemberIdq
  ) {
    final FeedDetailResponse feedResponse = feedService.findFeed(id, jwtMemberIdq.memberId());
    return BaseResponse.ok(feedResponse);
  }

  @PostMapping("/{id}/save")
  public BaseResponse<String> bookmarkFeed(
      final JwtMemberId jwtMemberId, @PathVariable final Long id
  ) {
    feedService.bookmarkFeed(jwtMemberId.memberId(), id);
    return BaseResponse.ok("");
  }

  @GetMapping("/save")
  public BaseResponse<List<FeedListResponse>> getBookmarkFeeds(final JwtMemberId jwtMemberId) {
    final var feeds = feedService.findBookmarkedFeeds(jwtMemberId.memberId());
    final List<FeedListResponse> responses = FeedListResponse.createList(feeds);
    return BaseResponse.ok(responses);
  }
}

File: src/main/kotlin/play/pluv/feed/controller/FeedController.kt
package play.pluv.feed.controller

import org.springframework.web.bind.annotation.*
import play.pluv.base.BaseResponse
import play.pluv.base.BaseResponse.Companion.ok
import play.pluv.feed.application.FeedService
import play.pluv.feed.application.dto.FeedDetailResponse
import play.pluv.feed.application.dto.FeedListResponse
import play.pluv.security.JwtMemberId

@RestController
@RequestMapping("/feed")
class FeedController(
    private val feedService: FeedService
) {

    @get:GetMapping
    val feeds: BaseResponse<List<FeedListResponse>>
        get() {
            val feeds = feedService.findAll()
            val responses = FeedListResponse.createList(feeds)
            return ok(responses)
        }

    @GetMapping("/{id}")
    fun getFeed(
        @PathVariable id: Long, jwtMemberIdq: JwtMemberId
    ): BaseResponse<FeedDetailResponse> {
        val feedResponse = feedService.findFeed(id = id, memberId = jwtMemberIdq.memberId)
        return ok(feedResponse)
    }

    @PostMapping("/{id}/save")
    fun bookmarkFeed(
        jwtMemberId: JwtMemberId, @PathVariable id: Long
    ): BaseResponse<String> {
        feedService.bookmarkFeed(memberId = jwtMemberId.memberId, feedId = id)
        return ok("")
    }

    @DeleteMapping("/{id}/save")
    fun cancelBookmarkFeed(
        jwtMemberId: JwtMemberId, @PathVariable id: Long
    ): BaseResponse<String> {
        feedService.cancelBookmark(jwtMemberId.memberId, id)
        return ok("")
    }

    @GetMapping("/save")
    fun getBookmarkFeeds(jwtMemberId: JwtMemberId): BaseResponse<List<FeedListResponse>> {
        val feeds = feedService.findBookmarkedFeeds(memberId = jwtMemberId.memberId)
        val responses = FeedListResponse.createList(feeds)
        return ok(responses)
    }
}
